### ISSUE TITLE
class imbalance, problem formulation

### ISSUE THREAD

**Issue by meera26** [fchollet/keras#4247]
hello

(binary classification)
i have 40,000 samples.
for each sample, input is a 2000 dimension feature vector. each feature has value either 0 or 1.
for each sample, output is one class label either 0 or 1.

i used a simple regression type model
##############################
model=Sequential()
model.add(Dense(1,input_dim=2000,init='normal',activation='sigmoid'))
model.compile(loss='binary_crossentropy',optimizer=SGD(lr=0.0001))
model.fit(xtrain,ytrain,batch_size=32,nb_epoch=80,verbose=1,validation_data=(xval,yval))
##############################

the problem is that among 40,000 samples, very few (5 to 10) samples have class label=1. (class imbalance)

I still want to train a model. However, i am new to keras and i am unable to formulate the problem.
Approach - using batch size = 2. for each batch select one sample with class label 0 and other sample with class label 1. There would be repetitive use of samples with class label 1 but that is okay.

Q1. is this handled by using class_weight in model.fit? Giving more weight to class label1 and less weight to class label0
Q2. is this handled by using sample_weight in model.fit? giving more weight to samples with class label1?
Q3. how do i make the model select a particular sample(i want one with class label1 and other with class label0) in a batch size?
Q4. or shall i do resampling outside keras? however it would take a lot of memory ?

**Comment by bityangke** [comment]
Hi @meera26 
Have you solved this problem？I have the same situation as you.How did you solve this at last?

**Comment by meera26** [comment]
Hi

There were a couple of solutions suggested ->
https://groups.google.com/forum/#!topic/keras-users/DqTUVKhstV8


There were two approaches that I tried -

The main issue is there are not enough 1s and the model could not learn
when to give a label 1 and so usually ends up predicting a large number of
1s.

*Approach 1* -
Using the custom loss function to reduce false positives.
lambda1 > lambda2.
For me choice of lambda1 and lambda2 as 1000 and 1 respectively worked fine.
##############################
def custom_loss(ytrue, ypred):
     reg1=-lambda1*ytrue*K.log(ypred)
     reg2=-lambda2*(1.0-ytrue)*K.log(1.0-ypred)
     lossvalue=K.mean(reg1+reg2)
     return lossvalue
##############################
model=Sequential()
model.add(Dense(1,input_dim=2000,init='normal',activation='sigmoid'))
model.compile(loss=custom_loss,optimizer=SGD(lr=0.0001))
model.fit(xtrain,ytrain,batch_size=32,nb_epoch=80,verbose=1,validation_data=(xval,yval))
##############################
ytest_pred=model.predict(xtest)
ytest_pred[ytest_pred>=0.5]=1
ytest_pred[ytest_pred<0.5]=0
##############################


*Approach 2 - *
Train 2000 separate models one for each class.
However, this assumes the independence relation between the 2000 classes.
In this we can incorporate the re-sampling thing so that the model sees
enough #1s and #0s.


for x in range(0,2000):
    #####################
    model=Sequential()
    model.add(Dense(1,input_dim=2000,init='normal',activation='sigmoid'))
    print(model.summary())
    batch_size=2
    nb_epoch=30
    opti=SGD(0.0001)
    model.compile(loss='binary_crossentropy',optimizer=opti)
    es=EarlyStopping(monitor='val_loss', patience=10, verbose=0,
mode='auto')
    #####################
    # the following prepares data with alternating 0 and 1 label samples #
    ydata1=ytrain[:,x]
    zero=[]
    one=[]

    ydata1=ydata1.tolist()
    for x in range(len(ydata1)):
        if ydata1[x]==0:
            zero.append(x)
        elif ydata1[x]==1:
            one.append(x)

    ydata=[]
    xdata=[]
    print(len(zero),len(one))
    for x in range(len(zero)):
            xdata.append(xtrain[zero[x]])
            ydata.append(ytrain[zero[x]])
            if len(one)!=0:
                i=random.randint(0,len(one)-1)
                xdata.append(xtrain[one[i]])
                ydata.append(ytrain[one[i]])

    xdata=np.array(xdata)
    ydata=np.array(ydata)
    print(xdata.shape, ydata.shape)
    #####################
    history=model.fit(xdata, ydata,batch_size=batch_size,
nb_epoch=nb_epoch,verbose=1,validation_split=0.3,callbacks=[es])
    predstest=model.predict(xtest)
    cPickle.dump(predstest,open('pred_'+str(x)+'.p','wb'))

so in the end the for-loop, there would be predictions for each of 2000
classes. Stack them up together and check for accuracy against groundtruth.

This was very crude way of resampling. There would be better ways to do it.

Approach 2 worked better than approach 1 for me. However, as stated in the
forum (above link), the class imbalance is actually very huge. However,
doing some manual post-processing to reduced the false positives can be
done.

I m still not happy with it, and looking if the problem can be formulated
differently to avoid so many 0s.
I am open to suggestions to tackle this problem.

Thanking you
Regards
